lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -0,0 +1,202 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+-- |
+-- Copyright: © 2020 IOHK
+-- License: Apache-2.0
+--
+-- Haskell-node "shelley" implementation of the @StakePoolLayer@ abstraction,
+-- i.e. some boring glue.
+module Cardano.Wallet.Shelley.Pools where
+
+import Prelude
+
+import Cardano.Wallet.Api.Server
+    ( LiftHandler (..), apiError )
+import Cardano.Wallet.Api.Types
+    ( ApiErrorCode (..), ApiT (..) )
+import Cardano.Wallet.Network
+    ( NetworkLayer (..) )
+import Cardano.Wallet.Primitive.Types
+    ( Coin (..), GenesisParameters (..), PoolId )
+import Cardano.Wallet.Shelley.Compatibility
+    ( Shelley
+    , ShelleyBlock
+    , fromPoolDistr
+    , fromRewards
+    , optimumNumberOfPools
+    , toPoint
+    , toShelleyCoin
+    )
+import Cardano.Wallet.Shelley.Network
+    ( pattern Cursor )
+import Cardano.Wallet.Unsafe
+    ( unsafeMkPercentage, unsafeRunExceptT )
+import Control.Monad.Class.MonadSTM
+    ( MonadSTM, TQueue )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Control.Monad.Trans.Except
+    ( ExceptT (..), runExceptT, withExceptT )
+import Data.Map
+    ( Map )
+import Data.Map.Merge.Strict
+    ( dropMissing, traverseMissing, zipWithMatched )
+import Data.Ord
+    ( Down (..) )
+import Data.Quantity
+    ( Percentage (..), Quantity (..) )
+import Data.Sort
+    ( sortOn )
+import Data.Word
+    ( Word64 )
+import GHC.Generics
+    ( Generic )
+import Ouroboros.Network.Block
+    ( Point )
+import Ouroboros.Network.Client.Wallet
+    ( LocalStateQueryCmd (..), send )
+import Servant
+    ( Handler, err500 )
+
+import qualified Cardano.Wallet.Api.Types as Api
+import qualified Data.Map.Merge.Strict as Map
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import qualified Ouroboros.Consensus.Shelley.Ledger as OC
+
+-- | Stake Pool Data fields fetched from the node via LSQ
+data PoolLsqMetrics = PoolLsqMetrics
+    { nonMyopicMemberRewards :: Quantity "lovelace" Word64
+    , stake :: Percentage
+    , saturation :: Double
+    } deriving (Eq, Show, Generic)
+
+
+data ErrFetchMetrics = ErrFetchMetrics
+  deriving Show
+
+askNode
+    :: MonadSTM m
+    => TQueue m (LocalStateQueryCmd ShelleyBlock m)
+    -> Point ShelleyBlock
+    -> Coin
+    -> ExceptT ErrFetchMetrics m (Map PoolId PoolLsqMetrics)
+askNode queue pt coin = do
```

Comment:
It's good to avoid _too general_ verbs like "ask" , "do", "run" etc .. in function names because they are often meaningless. `fetchMetrics` sounds a bit more transparent.

Reply:
b65293b4c54fc298b10f2adcfdcde61bc5dec4e1

Went with `fetchLsqPoolMetrics`. Want to highlight that this just fetches part of the data we want with the `Lsq` part.